src/server/req_runjob.c

```diff
@@ -802,14 +807,20 @@ req_runjob2(struct batch_request *preq, job *pjob)
 	rq_type = preq->rq_type;
 	if (preq && (rq_type == PBS_BATCH_AsyrunJob_ack)) {
 		reply_ack(preq);
-		preq = 0;	/* cleared so we don't try to reuse */
+		preq = NULL;	/* cleared so we don't try to reuse */
 	}
 
 	if (((rc = svr_startjob(pjob, preq)) != 0) &&
 		((rq_type == PBS_BATCH_AsyrunJob_ack) || preq)) {
+		save_subjob(pjob);
 		free_nodes(pjob);
-		if (preq)
-			req_reject(rc, 0, preq);
+		req_reject(rc, 0, preq);
+	}
+
+	/* if a subjob, the job wouldn't have gotten saved yet, let's save it now */
+	if (save_subjob(pjob)) {
```

Comment:
u seem to have just shifted the location of the subjob save, how is it cutting a job save. Certainly the save inside of svr_setjobstate() is also still happening?

Reply:
svr_setjobstate() doesn't save to db if newobj is set, which it is for the subjob. Calling job_save_db() before svr_setjobstate() clears that bit and causes svr_setjobstate() to call job_sabe_db again. That's how moving the job_save_db call after svr_setjobstate keeps the bit set and svr_setjobstate doesn't call job_save_db. I've added this explanation to the PR description now.

Reply:
Thanks for explaining, makes sense.

Reply:
Could you add a comment explaining why we are not saving the subjob before and how setjobstate() will not save it - somebody else will certainly forget this detail (i myself added the check for newjobstate in setjobstate() but did not recall that)